### PR TITLE
unicode support

### PR DESCRIPTION
**Pull request by obfusk**
Patch mentioned in #33 (but for `master`).

### CODE REVIEW

**Comment by obfusk** [comment]
I managed to get rid of the `FIXME`s and it seems to work quite well :smile:
It's still a bit rough though.

**Comment by mattip** [comment]
Does this work with python2/python3?

**Comment by mattip** [comment]
I guess so, tests are passing

**Comment by obfusk** [comment]
> Does this work with python2/python3?

Yes. The tests originally failed with python2, but I fixed that.
I've now tested it myself with pypy, pypy3, and cpython3.9.

**Comment by obfusk** [comment]
It doesn't handle e.g. zero width joiners, though stand-alone emoji work fine.
But `readline` doesn't really handle those either, and neither do many terminals AFAIK.

**Comment by blueyed** [comment]
Thanks for this already!
I wonder how good test coverage in this regard really is, and think that it would be great to get e.g. the corner cases you've fixed covered.

Unfortunately codecov is acting up, at least with the latest commits, since it would be useful to revisit/see its patch status (i.e. how much / which code is covered in tests after all).

**Comment by obfusk** [comment]
>  I wonder how good test coverage in this regard really is, and think that it would be great to get e.g. the corner cases you've fixed covered.

The tests don't really seem to cover any of the curses stuff. Which is of course hard to test.
At least the tests helped me fix the python2 compatibility.

I think I may have fixed one or two existing bugs. But I may also have introduced some :sweat_smile:

**Comment by blueyed** [comment]
(closed/re-opened to get codecov's status then hopefully)

File: pyrepl/reader.py [hunk 1]
```diff
@@ -20,15 +20,27 @@
...
     unicode = str
     unichr = chr
     basestring = bytes, str
+    decode = lambda x, _ = None: x
+
+
+def width(c):
+    return 2 if unicodedata.east_asian_width(c) in "FW" else 1
+def wlen(s):
+    return sum(map(width, s))
```
Comment (blueyed) [inline pyrepl/reader.py:43]: What do you think about using https://pypi.org/project/wcwidth/?

**Review by blueyed — COMMENTED** [review]
(Manual codecov link: https://app.codecov.io/gh/pypy/pyrepl/compare/34/diff)

File: pyrepl/reader.py (hunk 1 above)
Comment (mattip) [reply pyrepl/reader.py:43]: Looking in from the PyPy side, wcwidth seems like a small package with JIT-freindly structure (no complicated classes or deep function chains). It was dormant from 2016 to 2020, but seems to be active again. The license is MIT, which is fine for PyPy.

Comment (obfusk) [reply pyrepl/reader.py:43]: I agree that that bit of code could be improved, but what would be the advantage of using `wcwidth` over python's built-in unicode support? It seems to provide mostly the same functionality except with its own unicode tables that need to be kept up to date.

**Comment by obfusk** [comment]
> (Manual codecov link: https://app.codecov.io/gh/pypy/pyrepl/compare/34/diff)

That link gives me a 403.

File: pyrepl/reader.py (hunk 1 above)
Comment (obfusk) [reply pyrepl/reader.py:43]: `wcwidth` can also return `0` or `-1`, which the current code doesn't know how to handle. And I'm not sure what the best way to handle those would be tbh.

For improved performance, we could use e.g. `@functools.lru_cache`.

Comment (obfusk) [reply pyrepl/reader.py:43]: (note that `pyrepl` already shows e.g. control characters as `^c` etc.)

Comment (obfusk) [reply pyrepl/reader.py:43]: I'm not entirely sure about unicode combining characters. But normalisation before printing might be the better approach there.

Comment (obfusk) [reply pyrepl/reader.py:43]: Of course I could be missing something. Clearly `wcwidth` handles a lot more cases than my two lines of code.

Comment (blueyed) [reply pyrepl/reader.py:43]: /cc @jquast

Comment (jquast) [reply pyrepl/reader.py:43]: I keep wcwidth active, and happily accept improvements that I think go beyond this technique by a lot. I do plan a new release soon to better support more edge cases than the ones given, and open issues and such spell out any deficiencies well. 

By the way I’m not sure normalization matters, in the case of how wcswidth does accounting, the math works the same, but I could be wrong!

And control characters and such as displayed you’ll have to manage as they are displayed rather than their value. The wcwidth and wcswidth functions are a bit kludgey, thanks posix? We need to guard against -1 return value, and I hope to release new “width” function next release that I think is more pythonic

Comment (jquast) [reply pyrepl/reader.py:43]: Also to deal with -1, use max(result, 0), and, use wcwidth for ea. Character,

https://github.com/jquast/blessed/blob/master/blessed/sequences.py#L345

Comment (obfusk) [reply pyrepl/reader.py:43]: > By the way I’m not sure normalization matters, in the case of how wcswidth does accounting, the math works the same, but I could be wrong!

The normalisation comment wasn't directed at wcswidth but at my code (which currently doesn't handle combining characters etc.)

Comment (obfusk) [reply pyrepl/reader.py:43]: It looks like neither my code nor wcwidth (nor many terminals) handle emoji properly :sweat_smile: 

Screenshot from 2021-02-25 20-06-00 cleaned

Comment (jquast) [reply pyrepl/reader.py:43]: Yes, this case in particular will be in next wcwidth release, https://github.com/jquast/wcwidth/issues/39#issuecomment-769878110

Comment (obfusk) [reply pyrepl/reader.py:43]: Dealing with `-1` is indeed trivial. I would have to take a closer look at the code to make sure `0` is handled properly (since it was written based on the assumption `width()` only returns `1` or `2` and control characters are handled elsewhere).

If the upcoming new version of wcwidth handles emoji better, using it would be an improvement indeed.

... except that not every terminal would actually display emoji with the expected width :(

Comment (jquast) [reply pyrepl/reader.py:43]: it is possible to find the Unicode version of a terminal by somewhat intrusive detection https://github.com/jquast/ucs-detect

a user can export UNICODE_VERSION for their terminal, and wcwidth library matches their expected rendered form.

Hopefully the emulators themselves in the future could export it. https://jeffquast.com/post/terminal_wcwidth_solution/

I tried using ucs-detect in a bash profile for a while but it swallows input so it is probably too intrusive for most use cases.

Comment (obfusk) [reply pyrepl/reader.py:43]: > a user can export UNICODE_VERSION for their terminal, and wcwidth library matches their expected rendered form.

Unicode version != emoji version though. I don't think this helps with libvte-based terminals (like lxterminal and gnome-terminal) rendering e.g. 🤦‍♀️ as 🤦♀️. Both characters are supported (thus the unicode version is high enough), but the combined form is not (thus the emoji version isn't).

Comment (obfusk) [reply pyrepl/reader.py:43]: At least that's how I think it works. I find it somewhat confusing tbh.

Comment (obfusk) [reply pyrepl/reader.py:43]: > it is possible to find the Unicode version of a terminal by somewhat intrusive detection https://github.com/jquast/ucs-detect

According to the table in the README, lxterminal and kitty support the same unicode version (11.0.0).
But as my screenshot shows, they don't display emoji the same way.

(Slightly OT: I'd consider switching to kitty, but unfortunately it doesn't support Japanese input :crying_cat_face:)

Comment (obfusk) [reply pyrepl/reader.py:43]: I briefly tested using wcwidth: https://github.com/obfusk/pyrepl-wide/commit/eba676abb1638fc83a49bbac1d208cb8c3794316.

* I don't notice any improvement atm.
* I had to modify the original prompt length calculation (because that subtracts the length of ansi escape sequences).
* `_my_unctrl()` currently turns a zero-width joiner into `"\u200d"` and I'm not sure returning it unmodified won't break something else.

Comment (mattip) [reply pyrepl/reader.py:43]: @blueyed is there a suspicion that this PR is worse in some cases than the current code? If not maybe it can be merged as-is and other alternatives could be parts of future PRs?

Comment (obfusk) [reply pyrepl/reader.py:43]: Even with wcwidth's upcoming `width()` function (and setting aside the differences between terminals), handling zero-width joiners etc. properly is not easy, since `width(s) == sum(width(c) for c in s)` is no longer true for all strings. This would presumably require more careful and extensive modifications to the code than the current PR.

I think wcwidth is likely to be quite useful in improving unicode support even more in the future, but for now I think this PR is a good first step. Further improvements get a lot more complicated and should probably be separate, future PRs.

I'd also point out that `readline` doesn't really seem to handle emoji any better.

AFAIK this PR doesn't make anything worse. Unfortunately (but understandably since I don't really know a good way to test all the curses stuff), the test suite can't rule that out.

I do however think the PR is still a little rough and could use a code review and probably some updates to comments and documentation before merging.

Comment (obfusk) [reply pyrepl/reader.py:43]: > Looking in from the PyPy side, wcwidth seems like a small package with JIT-freindly structure (no complicated classes or deep function chains). It was dormant from 2016 to 2020, but seems to be active again. The license is MIT, which is fine for PyPy.

I don't think wcwidth supports python2 though, which I presume is an issue for PyPy.

**Comment by mattip** [comment]
Any more thoughts here? I would like to get this in the upcoming PyPy release.

**Comment by obfusk** [comment]
> Any more thoughts here? I would like to get this in the upcoming PyPy release.

To summarise:

* This handles double-width characters as found in e.g. Japanese and Chinese reasonably well.
* It seems to be an improvement over the current lack of unicode support and AFAIK doesn't make anything worse than before.
* It doesn't handle emoji (but that's hard and neither does `readline` afaik).
* It doesn't seem to handle e.g. Arabic (but I don't know much about that), but I don't think it makes it worse either.



* I've tested it with Japanese and found no (further) bugs, but it would be nice to have a comprehensive test suite.
* I'd like someone to review the code to make sure I didn't miss anything.
* The PR is a bit rough and could probably use a bit of refactoring; some old comments should probably be updated.



* Future improvements would probably benefit from switching to `wcwidth`, but this will require more extensive modifications as some invariants/assumptions (such as `width(a) + width(b) == width(a+b)`) will no longer hold.
